src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -1019,7 +1019,18 @@ static QVariant fcnToString( const QVariantList &values, const QgsExpressionCont
 
 static QVariant fcnToDateTime( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
 {
-  return QVariant( QgsExpressionUtils::getDateTimeValue( values.at( 0 ), parent ) );
+  QString format = QgsExpressionUtils::getStringValue( values.at( 1 ), parent );
+  if ( format.isEmpty() )
+    return QVariant( QgsExpressionUtils::getDateTimeValue( values.at( 0 ), parent ) );
+
+  QString datetimestring = QgsExpressionUtils::getStringValue( values.at( 0 ), parent );
+  QDateTime datetime = QDateTime::fromString( datetimestring, format );
+  if ( !datetime.isValid() )
```

Comment:
What about empty string values? They shouldn't cause an evaluation error, but instead should return a null value without error.

Reply:
@nyalldawson , I think we should stick to current behavior (predating this PR) whereas all of our `to_*()` functions return an error when passing on an empty string. That's actually the reason why I added `try()` a while back.